lib/git-fastclone.rb

```diff
@@ -333,24 +338,26 @@ def update_reference_repo(url, fail_hard)
         submodule_file = reference_repo_submodule_file(url, reference_dir, using_local_repo)
 
         # if prefetch is on, then grab children immediately to frontload network requests
-        prefetch(submodule_file) if File.exist?(submodule_file) && prefetch_submodules
+        prefetch(submodule_file, attempt_number) if File.exist?(submodule_file) && prefetch_submodules
 
         # Store the fact that our repo has been updated if necessary
-        store_updated_repo(url, mirror, repo_name, fail_hard) unless reference_updated[repo_name]
+        store_updated_repo(url, mirror, repo_name, fail_hard, attempt_number) unless reference_updated[repo_name]
       end
     end
 
     # Grab the children in the event of a prefetch
-    def prefetch(submodule_file)
+    def prefetch(submodule_file, attempt_number)
       File.readlines(submodule_file).each do |line|
         # We don't join these threads explicitly
-        Thread.new { update_reference_repo(line.strip, false) }
+        Thread.new { update_reference_repo(line.strip, false, attempt_number) }
       end
     end
 
     # Creates or updates the mirror repo then stores an indication
     # that this repo has been updated on this run of fastclone
-    def store_updated_repo(url, mirror, repo_name, fail_hard)
+    def store_updated_repo(url, mirror, repo_name, fail_hard, attempt_number)
+      trigger_pre_clone_hook(url, mirror, attempt_number)
```

Comment:
```suggestion
      trigger_pre_clone_hook(url, mirror, attempt_number) unless unless Dir.exist?(mirror)
```

Note that we had such logic on line 453: we only invoke `popen2e_wrapper` if we are going to clone soon (because we are in the `store_updated_repo`). However, if a pre-clone-hook logic correctly performs a "patch"  and creates a directory that we are going to create with `git clone`, there is no need to make clone and instead continue with `git update`. Therefore we check  `unless Dir.exist?(mirror)` twice (line 361 and 453).